packages/mermaid/src/diagram-api/frontmatter.spec.ts

```diff
@@ -0,0 +1,71 @@
+import { vi } from 'vitest';
+import { extractFrontMatter } from './frontmatter';
+
+const dbMock = () => ({ setDiagramTitle: vi.fn() });
+
+describe('extractFrontmatter', () => {
+  it('returns text unchanged if no frontmatter', () => {
+    expect(extractFrontMatter('diagram', null)).toEqual('diagram');
+  });
+
+  it('returns text unchanged if frontmatter lacks closing delimiter', () => {
+    const text = `---\ntitle: foo\ndiagram`;
+    expect(extractFrontMatter(text, null)).toEqual(text);
+  });
+
+  it('handles empty frontmatter', () => {
+    const db = dbMock();
+    const text = `---\n\n---\ndiagram`;
+    expect(extractFrontMatter(text, db)).toEqual('diagram');
+    expect(db.setDiagramTitle).not.toHaveBeenCalled();
+  });
+
+  it('handles frontmatter without mappings', () => {
+    const db = dbMock();
+    const text = `---\n1\n---\ndiagram`;
+    expect(extractFrontMatter(text, db)).toEqual('diagram');
+    expect(db.setDiagramTitle).not.toHaveBeenCalled();
+  });
+
+  it('does not try to parse frontmatter at the end', () => {
+    const db = dbMock();
+    const text = `diagram\n---\ntitle: foo\n---\n`;
+    expect(extractFrontMatter(text, db)).toEqual(text);
+    expect(db.setDiagramTitle).not.toHaveBeenCalled();
+  });
+
+  it('handles frontmatter with multiple delimiters', () => {
+    const db = dbMock();
+    const text = `---\ntitle: foo---bar\n---\ndiagram\n---\ntest`;
+    expect(extractFrontMatter(text, db)).toEqual('diagram\n---\ntest');
+    expect(db.setDiagramTitle).toHaveBeenCalledWith('foo---bar');
+  });
```

Comment:
I don't see any way of supporting that syntax without serious compromises:
1. Disallowing leading whitespace before the delimiters would work (that's what gray-matter does), but leading whitespace before a diagram is very common, and disallowing it means you can't indent a diagram. e.g. https://github.com/mermaid-js/mermaid/blob/70f024735b57ac5b29b5c339ac871078ab46fa5a/demos/flowchart.html#L19-L24
2. Changing the `(.*?)` to `(.*)` isn't an option because that will break compatibility with diagrams containing `---`
3.  Trying to match the indent level of the closing delimiter with that of the opening delimiter won't work because leading whitespace is stripped in https://github.com/mermaid-js/mermaid/blob/70f024735b57ac5b29b5c339ac871078ab46fa5a/packages/mermaid/src/mermaid.ts#L149-L153

Do you have any other ideas? If not, I think we should consider this an edge case and not support it

Reply:
> Disallowing leading whitespace before the delimiters would work

My gut feeling is that this is the approach we should go for (as that's that is what [YAML spec v1.0 section 4.3.1](https://yaml.org/spec/1.0/#syntax-stream-doc) says), but you are right that many people do indent the diagrams.

We could add a [`dedent`](https://www.npmjs.com/package/ts-dedent) to https://github.com/mermaid-js/mermaid/blob/70f024735b57ac5b29b5c339ac871078ab46fa5a/packages/mermaid/src/mermaid.ts#LL149-L153. It would only apply to HTML however.

We could also add `dedent` inside `parse(str)` somewhere (or other functions that takes the Mermaid string directly), e.g. right before:

https://github.com/mermaid-js/mermaid/blob/09ed41b7d2cd5c8ffd50c7544bc88bd213d3c601/packages/mermaid/src/mermaidAPI.ts#L81

But that's not super required. People calling the JavaScript API can use `dedent` themselves if they want to use indented strings and the YAML frontmatter, e.g.

```js
import { dedent } from 'ts-dedent';
const myDiagram = dedent`
  ---
  title: This YAML is indented by two spaces.
  ---
  flowchart
    "This diagram is indented by two spaces."
`;

const svg = mermaid.render("mermaid-render-element", myDiagram);
```

Reply:
Well, alright, if you're sure that's the way to go, I implemented that in 7eba4f930bf56d8811ab00c93ad0b875dbd4c4c0

Reply:
> if you're sure that's the way to go

I'm not fully sure! But I'd rather have the parser be too strict at the the beginning, since we can always loosen things later without it being a breaking change. It's much harder to make something stricter later on, as that would be a breaking change.

But I'd be be happy for people to disagree or if they think there's a better way to do things!

I am slightly leaning towards also adding a `dedent` before parsing the Mermaid HTML in [mermaid.init/mermaid.initThrowsErrors](https://github.com/mermaid-js/mermaid/blob/70f024735b57ac5b29b5c339ac871078ab46fa5a/packages/mermaid/src/mermaid.ts#L149-L153) so that Mermaid diagrams indented in HTML work, but not for JS code in `mermaid.parse`.

But that could be a decision for a future PR :)

Reply:
When adding diagrams in html you inherit the indentation from the tag. This looks better in HTML:
```
    <pre id="diagram" class="mermaid">
      ---
      title: Example Git diagram
      ---
      gitGraph
        commit
        commit
    </pre>
```

than:
```
    <pre id="diagram" class="mermaid">
---
title: Example Git diagram
---
gitGraph
  commit
  commit
    </pre>
```

Even though it is less compliant with the standard.

I don't see the multi-line issue as a big problem though. My guess is that the user impact of indenting will affect far, far more users then the lack of ability to use `---` in the beginning of a a multiline row. 

I lean towards the more loose application, better to let the very few users that wants to have the --- in the multiline comment fail instead of irritating the users that want to indent.

But Alois has a fair point. Once we let the cat out of the box, it is out. Lets start with this but I suspect that we need to discuss this more and that we might might want to reverse the a11ab3d5ea1477c02101eb7c4279b1dc0d2ef126 commit. That can be outside this PR though. Good job @MasonM Solid!